extensions/wikia/CommunityPage/models/CommunityPageSpecialUsersModel.class.php

```diff
@@ -0,0 +1,371 @@
+<?php
+
+class CommunityPageSpecialUsersModel {
+	const TOP_CONTRIB_MCACHE_KEY = 'community_page_top_contrib';
+	const FIRST_REV_MCACHE_KEY = 'community_page_first_revision';
+	const GLOBAL_BOTS_MCACHE_KEY = 'community_page_global_bots';
+	const ALL_MEMBERS_MCACHE_KEY = 'community_page_all_members';
+	const MEMBER_COUNT_MCACHE_KEY = 'community_member_count';
+	const RECENTLY_JOINED_MCACHE_KEY = 'community_page_recently_joined';
+	const CURR_USER_CONTRIBUTIONS_MCACHE_KEY = 'community_page_current_user_contributions';
+
+	private $wikiService;
+
+	private function getWikiService() {
+		return $this->wikiService;
+	}
+
+	public function __construct() {
+		$this->wikiService = new WikiService();
+	}
+
+	/**
+	 * Get the date a member made their first edit to a wiki.
+	 * Membership is defined as having edited in the last two years
+	 *
+	 * @param User $user
+	 * @return Mixed|null
+	 */
+	public function getFirstRevisionDate( User $user ) {
+		if ( $user->isAnon() ) {
+			return null;
+		}
+
+		$data = WikiaDataAccess::cache(
+			wfMemcKey( self::FIRST_REV_MCACHE_KEY ),
+			WikiaResponse::CACHE_STANDARD,
+			function () use ($user) {
+				$db = wfGetDB( DB_SLAVE );
+
+				$sqlData = ( new WikiaSQL() )
+					->SELECT( '*' )
+					->FROM ( 'wikia_user_properties' )
+					->WHERE ( 'wup_property' )->EQUAL_TO( 'firstContributionTimestamp' )
+					->AND_ ( 'wup_user' )->EQUAL_TO( $user->getID() )
+					->AND_ ( 'wup_value > DATE_SUB(now(), INTERVAL 2 YEAR)' )
+					->ORDER_BY( 'wup_value DESC' )
+					->run( $db, function ( ResultWrapper $result ) {
+						$out = [];
+						while ( $row = $result->fetchRow() ) {
+							$out[] = [
+								'wup_value' => $row['wup_value']
+							];
+						}
+
+						return $out;
+					} );
+
+				return $sqlData;
+			}
+		);
+
+		if (count( $data ) > 0) {
+			return $data[0]['wup_value'];
+		} else {
+			return null;
+		}
+	}
+
+	/**
+	 * Get the user id and contribution count of the top n contributors to the current wiki,
+	 * optionally filtered by admins only
+	 *
+	 * @param int $limit Number of rows to fetch
+	 * @param string $interval Time interval for DATE_SUB()
+	 * @param bool $onlyAdmins Whether to filter by admins
+	 * @return Mixed|null
+	 */
+	public static function getTopContributors( $limit = 10, $interval = '1 WEEK', $onlyAdmins = false ) {
+		$data = WikiaDataAccess::cache(
+			wfMemcKey( self::TOP_CONTRIB_MCACHE_KEY, $limit, $interval, $onlyAdmins ),
+			WikiaResponse::CACHE_STANDARD,
+			function () use ( $limit, $interval, $onlyAdmins ) {
+				global $wgExternalSharedDB;
+				$db = wfGetDB( DB_SLAVE );
+				$adminFilter = '';
+				if ( $onlyAdmins ) {
+					$adminFilter = ' AND (ug_group = "sysop")';
+				}
+
+				$sqlData = ( new WikiaSQL() )
+					->SELECT( 'user_name, user_id, count(rev_id) AS revision_count' )
+					->FROM ( 'revision FORCE INDEX (user_timestamp)' )
```

Comment:
What was the reason for calculating this with the local revision table, rather than using the rollups in the datawarehouse that calculate user edit counts for the last week, month, etc.?